spec/app/ics-101-interchain-swap/README.md

```diff
@@ -0,0 +1,1613 @@
+  abortTransactionUnless(found);
+
+  const order: MultiAssetDepositOrder = {
+    poolId: msg.poolId,
+    chainId: pool.sourceChainId,
+    sourceMaker: msg.deposits[0].sender,
+    destinationTaker: msg.deposits[1].sender,
+    deposits: getCoinsFromDepositAssets(msg.deposits),
+    status: "PENDING",
+    createdAt: store.blockTime(),
+  };
+
+  store.appendMultiDepositOrder(msg.poolId, order);
+
+  return {
+    poolTokens: stateChange.poolTokens,
+  };
+}
+
+function onTakeMultiAssetDepositReceived(
+  msg: MsgTakeMultiAssetDepositRequest,
+  stateChange: StateChange
+): MsgMultiAssetDepositResponse {
+  abortTransactionUnless(msg.validateBasic() === undefined);
+
+  const { pool, found } = store.getInterchainLiquidityPool(msg.poolId);
+  abortTransactionUnless(found);
+
+  const { order, found: orderFound } = store.getMultiDepositOrder(msg.poolId, msg.orderId);
+  abortTransactionUnless(orderFound);
+  order.status = "COMPLETE";
+
+  for (const supply of stateChange.poolTokens) {
+    pool.addPoolSupply(supply);
+  }
+
+  for (const asset of order.deposits) {
+    pool.addAsset(asset);
+  }
+
+  const totalPoolToken = sdk.NewCoin(msg.poolId, sdk.NewInt(0));
+  for (const poolToken of stateChange.poolTokens) {
+    totalPoolToken.amount = totalPoolToken.amount.add(poolToken.amount);
+  }
+
+  store.mintTokens(order.sourceMaker, totalPoolToken);
+
+  store.setInterchainLiquidityPool(pool);
+  store.setMultiDepositOrder(pool.id, order);
+
+  return {};
+}
+
+function onMultiAssetWithdrawReceived(
+  msg: MsgMultiAssetWithdrawRequest,
+  stateChange: StateChange
+): MsgMultiAssetWithdrawResponse {
+  abortTransactionUnless(msg.validateBasic() === undefined);
+  const { pool, found } = store.getInterchainLiquidityPool(msg.poolToken.denom);
+  abortTransactionUnless(found);
+
+  for (const poolAsset of stateChange.out) {
+    pool.subtractAsset(poolAsset);
+  }
+
+  for (const poolToken of stateChange.poolTokens) {
+    pool.subtractPoolSupply(poolToken);
+  }
+
+  store.unlockTokens(
+    pool.counterPartyPort,
+    pool.counterPartyChannel,
+    msg.counterPartyReceiver,
+    sdk.NewCoins(stateChange.out[1])
+  );
+
+  if (pool.supply.amount == 0) {
+    store.removeInterchainLiquidityPool(pool.id);
+  } else {
+    store.setInterchainLiquidityPool(pool);
+  }
+
+  return { tokens: stateChange.out };
+}
+
+function onSwapReceived(msg: MsgSwapRequest, stateChange: StateChange): MsgSwapResponse {
+  abortTransactionUnless(msg.validateBasic() === undefined);
+
+  const { pool, found } = store.getInterchainLiquidityPool(msg.poolId);
+  abortTransactionUnless(found);
+  store.unlockTokens(pool.counterPartyPort, pool.counterPartyChannel, msg.recipient, sdk.NewCoins(stateChange.out[0]));
+
+  pool.subtractAsset(stateChange.out[0]);
+  pool.addAsset(msg.tokenIn);
+
+  store.setInterchainLiquidityPool(pool);
+
+  return { tokens: stateChange.out };
+}
+
+function onMakePoolAcknowledged(msg: MsgMakePoolRequest, poolId: string): void {
```

Comment:
If the make pool flow completes successfully initiated from chain A, but there is not take pool executed from the counterparty chain B. Is there a way for the creator of the pool on chain A, to recover their tokens?

Reply:
yes we've adding cancel logic.